### ISSUE TITLE
Font size is extra extra extra large

### ISSUE THREAD

**Issue by rlue** [jwilm/alacritty#1501]
Just compiled alacritty on Debian testing (buster), and when I opened it, the text was jumbo-sized. I have a laptop running the exact same OS where alacritty works perfectly, so I compiled against an older version, and it rendered just fine.

I went through the git log history, compiling various versions of alacritty until I found the culprit commit: 765949f.

Below is a screenshot of 765949f on the left and its parent (f7172fd) on the right:

![alacritty-comparison](https://user-images.githubusercontent.com/12194123/43883379-82c60546-9be5-11e8-818e-e6ba4821bffe.png)

These are, of course, on the same `alacritty.yml` config file, where `font:` has `size: 13.0`.

On KDE 5 / X11, FYI.

**Comment by chrisduerr** [comment]
This might just be DPI scaling taking effect. Do you have any DPI configured in X11 at all?

**Comment by rlue** [comment]
Nope!

**Comment by chrisduerr** [comment]
Is this on a single screen or do you have multiple displays connected?

**Comment by chrisduerr** [comment]
Any warnings/errors when you start Alacritty from a terminal emulator so you can see the stderr/out?

Also please post the full configuration you're using.

**Comment by rlue** [comment]
Sorry, no errors:

![screenshot_20180813_170128](https://user-images.githubusercontent.com/12194123/44022584-9fc602da-9f1a-11e8-82e8-6231de8d5215.png)

My config is [here](https://github.com/rlue/.config/blob/master/alacritty/alacritty-hybrid.yml).

**Comment by chrisduerr** [comment]
Could you give me the output of running Alacritty with the command line flags `alacritty -vvv`?

This should print all necessary debug info.

**Comment by rlue** [comment]
#### Against `master`

```
Welcome to Alacritty.
Configuration loaded from /home/rlue/.config/alacritty/alacritty.yml
device_pixel_ratio: 1.6666666
width: 1024, height: 768
Initializing glyph cache
Finished initializing glyph cache in 0.109033465
width: 1330, height: 820
Cell Size: (15 x 34)
PTY Dimensions: Line(24) x Column(88)
Scaling RGB by 0.66 from Rgb { r: 208, g: 210, b: 209 } to Rgb { r: 137, g: 138, b: 137 }
Using config-provided dim colors
Scaling RGB by 0.66 from Rgb { r: 208, g: 210, b: 209 } to Rgb { r: 137, g: 138, b: 137 }
Using config-provided dim colors
Font size changed: Size(26)
Term::resize
num_cols, num_lines = 52, 14
width: 798, height: 492
Set Attribute: Foreground(Indexed(187))
Set Attribute: Reset
Set Attribute: Foreground(Indexed(174))
Set Attribute: Reset
```

#### Against f7172fd

```
Welcome to Alacritty.
Configuration loaded from /home/rlue/.config/alacritty/alacritty.yml
device_pixel_ratio: 1
width: 1024, height: 768
Initializing glyph cache
Finished initializing glyph cache in 0.158548293
width: 802, height: 508
Cell Size: (9 x 21)
PTY Dimensions: Line(24) x Column(88)
Using config-provided dim colors
Using config-provided dim colors
Font size changed: Size(26)
Term::resize
Term::resize dimensions unchanged
width: 802, height: 508
Set Attribute: Foreground(Indexed(187))
Set Attribute: Reset
Set Attribute: Foreground(Indexed(174))
Set Attribute: Reset
```

---

Again, I'm on KDE Plasma 5 / Debian testing (buster). Below are screenshots of my display scaling system preferences panels:

![screenshot_20180814_231013](https://user-images.githubusercontent.com/12194123/44100314-689e2c1c-a017-11e8-8a86-6f8726ab5681.png)

![screenshot_20180814_231036](https://user-images.githubusercontent.com/12194123/44100323-6beb5e94-a017-11e8-9c32-66be398fc786.png)

**Comment by chrisduerr** [comment]
You could try running #1403, this introduces a new major glutin update which potentially fixes this issue.

**Comment by rlue** [comment]
Sorry, no luck. :\

**Comment by chrisduerr** [comment]
Ooooh, I found the issue. I didn't notice it since in your `-vvv` log you posted the master first and the working branch second. I've looked at the `device_pixel_ratio` of the second output and it's 1.0, so there's no scaling.

However on the master branch you can see that the device_pixel_ratio (short dpr) is 1.6666666, so your font size 13 is scaled by 1.66666 (21.67). So if you change master to the font size `22`, it's will probably give you the same font size.

This behavior is intentional since the latest version of the windowing library will always automatically use DPR. So the font size you've specified (13) is in `pt`, which means that no matter the size or resolution of the display you open Alacritty on it will always be the same size (in mm).

Now there are two possible solutions you have to solve this problem right now. One is to decrease the font size. If you switch the font size to `8` you should get the same size you've had before the DPI change. Another option which is only available on X11 is to start Alacritty with the environment flag `WINIT_HIDPI_FACTOR=1.0 alacritty`. This will set a fixed DPI value of 1.0 for all your displays and should restore the behavior you've had previously. You could also export this in your shell's profile file or create an alias to make this permanent.

I could see that in the future it might be possible to override the DPI for all platforms in the Alacritty config, but since you're using X11 what is in place already should be fine to do the job for you.

I hope this solves this issue for you. I'm going to close it since it's not an actual issue with Alacritty (or the way this works is intentional). However feel free to ask more questions if you need help to get this working.

**Comment by rlue** [comment]
Sorry for the delay in getting back to you. I'm more than fine with this answer, but am a little puzzled on this point:

> This behavior is intentional since the latest version of the windowing library will always automatically use DPR. So the font size you've specified (13) is in `pt`, which means that no matter the size or resolution of the display you open Alacritty on it will always be the same size (in mm).

The bug I filed was occurring on both a 14" laptop display and a 23" desktop LCD monitor of the same resolution (1080p). If I'm understanding you correctly, the windowing library using DPR should mean that the actual size of the text on both displays is equivalent (or at least comparable) despite their very different DPIs, but in practice, it's jumbo-sized on both.

Is that not how it's supposed to work?

**Comment by chrisduerr** [comment]
The goal is to have the same size on both monitors, not in pixels, but in millimeters. Due to rounding this will never be 100% accurate, but it should be about the same size.

However there currently is no resizing when an Alacritty window changes the screen, so if it's scaled up on one screen and you move it to another where the scaling is `1.0`, it will be way too big. This is fixed with the #1403 PR and you might be able to make things work better by running that.

If you have two screens where you previously didn't use any DPI and they are now both properly recognized as `2.0` DPI, then the font size should be doubled now. This is how DPI is supposed to work even though a lot of applications on Linux don't make use of DPI (which is why the workaround I've mentioned before exists). If you like proper DPI scaling between your monitors, the proper way to 'fix' big font sizes is to reduce the font size in the config.

Hope that answers your question. If you have any more questions, please let me know.

**Comment by rlue** [comment]
I hear ya, and that sounds great. Unfortunately, it's working backwards on my computers:

### Desktop w/ 24" LCD display @ 1920x1080 resolution

```
$ alacritty -vvv
Welcome to Alacritty.
Configuration loaded from /home/rlue/.config/alacritty/alacritty.yml
device_pixel_ratio: 3.1666667
width: 1024, height: 768
Initializing glyph cache
Finished initializing glyph cache in 0.026040116
width: 2474, height: 1516
Cell Size: (28 x 63)
PTY Dimensions: Line(24) x Column(88)
Scaling RGB by 0.66 from Rgb { r: 208, g: 210, b: 209 } to Rgb { r: 137, g: 138, b: 137 }
Using config-provided dim colors
Scaling RGB by 0.66 from Rgb { r: 208, g: 210, b: 209 } to Rgb { r: 137, g: 138, b: 137 }
Using config-provided dim colors
Font size changed: Size(26)
Term::resize
num_cols, num_lines = 27, 7
width: 781, height: 478
Set Attribute: Foreground(Indexed(187))
Set Attribute: Reset
Set Attribute: Foreground(Indexed(174))
clear_line: Right
wrapping
carriage_return
Set Attribute: Reset
carriage_return
linefeed
Goodbye.
```

### Laptop with 14" 'hi-dpi' dissplay @ 1920x1080 resolution

```
$ alacritty -vvv
Welcome to Alacritty.
Configuration loaded from /home/rlue/.config/alacritty/alacritty.yml
device_pixel_ratio: 1.6666666
width: 1024, height: 768
Initializing glyph cache
Finished initializing glyph cache in 0.099498874
width: 1330, height: 820
Cell Size: (15 x 34)
PTY Dimensions: Line(24) x Column(88)
Scaling RGB by 0.66 from Rgb { r: 208, g: 210, b: 209 } to Rgb { r: 137, g: 138, b: 137 }
Using config-provided dim colors
Scaling RGB by 0.66 from Rgb { r: 208, g: 210, b: 209 } to Rgb { r: 137, g: 138, b: 137 }
Using config-provided dim colors
Font size changed: Size(26)
Term::resize
num_cols, num_lines = 52, 14
width: 798, height: 492
Set Attribute: Foreground(Indexed(187))
Set Attribute: Reset
Set Attribute: Foreground(Indexed(174))
Set Attribute: Reset
carriage_return
linefeed
Goodbye.
```

This is what they look like side-by-side:

![dsc_0015](https://user-images.githubusercontent.com/12194123/44449947-2b476c00-a622-11e8-9eba-78cf8b72654f.JPG)

---

Again, I'm totally happy currently building against an older commit, but it sounds like the current behavior is not what's intended, so I'm submitting this information as a bug report.

**Comment by chrisduerr** [comment]
@rlue I don't think a 24'' screen and 1920 x 1080 should ever be 3.1666667 DPR. Just for reference the usual hidpi screen of macbooks is a DPR of 2, this isn't even 1440p.

Could you share the `xrandr` output of that screen? It is possible that `xrandr` is reporting incorrect dimensions. The important parts are the resolution (1920x1080 hopefully) and the size of the screen in mm. Chances are if the size in `mm` is like < 100mm, there's something wrong.

**Comment by rlue** [comment]
```
$ xrandr
Screen 0: minimum 320 x 200, current 1920 x 1080, maximum 8192 x 8192
HDMI-1 connected primary 1920x1080+0+0 (normal left inverted right x axis y axis) 160mm x 90mm
   1920x1080     60.00*+  50.00    59.94
   1920x1080i    60.00    50.00    59.94
   1600x1200     60.00
   1680x1050     59.88
   1280x1024     60.02
   1440x900      59.90
   1280x960      60.00
   1280x800      59.91
   1280x720      60.00    50.00    59.94
   1024x768      60.00
   800x600       60.32    56.25
   720x576       50.00
   720x480       60.00    59.94
   640x480       60.00    59.94
DP-1 disconnected (normal left inverted right x axis y axis)
HDMI-2 disconnected (normal left inverted right x axis y axis)
```

`160mm x 90mm` -> I guess this is a bug in RandR?

**Comment by chrisduerr** [comment]
@rlue Yeah, this looks like a bug in RandR or an issue with your monitor. However, it is possible to fix these values in your X.Org config. I don't have it handy right now, so you're probably going to be as fast as me trying to look it up, but if you need some help with that, let me know.

Once the screen dimensions are set properly, the DPI should change to something more reasonable like `1.X` and this bug should resolve itself! Please report back if you were able to make it work. :)

**Comment by rlue** [comment]
I have not been able to make it work (though tbh I haven't tried _that_ hard). I don't have an Xorg config, and while I understand that [it would be pretty trivial to generate one](https://wiki.debian.org/Xorg#What_if_I_do_not_have_a_xorg_config_file.3F), I found a thread on the ArchLinux BBS that suggested that

> Xorg appears to ignore the **DisplaySize** of a "Monitor" section in a Xorg configuration file (see below). This solution would be sub-optimal anyway, because it would force all external HDMI monitors to report the same physical size. I would prefer to configure the monitor via xrandr if possible.

The OP of that thread said he got it working by extracting and modifying his monitor's EDID file (ostensibly with the `read-edid` package and `hexedit`). My monitor is 10 years old and `read-edid` reported the following error:

```
EDID claims 1 more blocks left
EDID blocks left is wrong.
Your EDID is probably invalid.
```

so I'm just gonna not worry about it anymore.

Thanks anyway for all your help!

**Comment by chrisduerr** [comment]
@rlue Note that if you have a desktop setup where monitors don't change frequently, you don't have to worry about something like EDID at all. That would just over-complicate things.

If you decide to just ignore DPI scaling, the `WINIT_HIDPI_FACTOR=1.0 alacritty` option should allow you to just ignore DPI scaling. But I really don't think that this should be much trouble to setup. Especially if you do not frequently switch between a variety of different monitors. You can also always just invoke `xrandr` manually, ignoring EDID completely and just running it when the right monitor is plugged in.

I think there are a variety of options to make this work. If EDID doesn't work out for you, there's still plenty of stuff to try.

**Comment by rlue** [comment]
Funny, when I change the DPI via `xrandr` (as below)

```
$ xrandr --dpi 90
$ xdpyinfo | grep dots
  resolution:    90x90 dots per inch
```

it changes the display scaling of _some_ new windows, but not all. For instance, the "System Settings", "Dolphin" (file browser), and "Okular" (document viewer) applications both get scaled accordingly, but Firefox, Chromium, and alacritty do not. Furthermore, alacritty's `device_pixel_ratio` does not change (still 3.16666667).

If it's not dependent on the DPI set by `xrandr`, then what setting needs to change?

**Comment by chrisduerr** [comment]
Alacritty ignores the DPI set through xrandr, since it doesn't allow for per display scaling.

The only thing Alacritty cares about is the physical size and resolution of the screen. It then calculates the DPI automatically to always be correct. This allows scaling windows based on the specific DPI of each monitor, instead of the DPI of the complete X screen.

**Comment by rawkode** [comment]
I'm experiencing this same problem. The font-size is only OK on one monitor. When configured at `size: 8`, it's OK on my laptop screen. When configured at `size: 12`, it's OK on my external display.

```
Screen 0: minimum 8 x 8, current 1920 x 1080, maximum 32767 x 32767
eDP1 connected primary (normal left inverted right x axis y axis)
   1920x1080     60.00 +  59.93    48.00
   1680x1050     59.88
   1400x1050     59.98
   1600x900      60.00    59.95    59.82
   1280x1024     60.02
   1400x900      59.96    59.88
   1280x960      60.00
   1368x768      60.00    59.88    59.85
   1280x800      59.81    59.91
   1280x720      59.86    60.00    59.74
   1024x768      60.00
   1024x576      60.00    59.90    59.82
   960x540       60.00    59.63    59.82
   800x600       60.32    56.25
   864x486       60.00    59.92    59.57
   640x480       59.94
   720x405       59.51    60.00    58.99
   640x360       59.84    59.32    60.00
DP1 connected 1920x1080+0+0 (normal left inverted right x axis y axis) 700mm x 390mm
   1920x1080    144.00 + 120.00*  119.88   120.00    60.00    50.00    59.94
   1920x1080i    60.00    50.00    59.94
   1680x1050     59.88
   1280x1024     75.02    60.02
   1440x900      59.90
   1280x960      60.00
   1280x800      59.91
   1152x864      75.00
   1280x720      60.00    50.00    59.94
   1024x768      75.03    70.07    60.00
   832x624       74.55
   800x600       72.19    75.00    60.32    56.25
   720x576       50.00
   720x480       60.00    59.94
   640x480       75.00    72.81    66.67    60.00    59.94
   720x400       70.08
DP2 disconnected (normal left inverted right x axis y axis)
HDMI1 disconnected (normal left inverted right x axis y axis)
HDMI2 disconnected (normal left inverted right x axis y axis)
VIRTUAL1 disconnected (normal left inverted right x axis y axis)
```

**Comment by gun1x** [comment]
> The only thing Alacritty cares about is the physical size and resolution of the screen. It then calculates the DPI automatically to always be correct. This allows scaling windows based on the specific DPI of each monitor, instead of the DPI of the complete X screen.

sounds great! however, my work monitor is 2k ultrawide 34 inches, and my laptop monitor is full hd 16:10 14 inches. This means that everything is great with `Xft.dpi: 117` in `.Xresources`. So if I set the font size to 7, it's ok for my laptop, but if I set the font size to 10, it's ok for my monitor.

is there any way to stop alacritty from having this behavior?

**Comment by chrisduerr** [comment]
Xft.dpi will always be ignored. However you can manually override the DPI with the `WINIT_HIDPI_FACTOR` environment variable.

Though it does sound like you don't actually wanat DPI scaling.

**Comment by gun1x** [comment]
> Though it does sound like you don't actually wanat DPI scaling.

Nope, I just want the font size to be pixel based, without any magic. Just normal font size, like in any other app.

Since i3 is really bad at handling monitors I have a script anyway to switch monitor, so I can make that script also sed the alacritty config to fix the font, if there is no other option.

**Comment by chrisduerr** [comment]
Then you should use `WINIT_HIDPI_FACTOR=1`.

**Comment by gun1x** [comment]
> Then you should use `WINIT_HIDPI_FACTOR=1`.

i added that to .xinitrc and it worked, thank you!

**Comment by elihunter173** [comment]
This might be a stupid question, but would it be possible for alacritty to have a setting in `alacritty.yml` to turn off DPI scaling?

**Comment by kchibisov** [comment]
@elihunter173 We had this setting in the past, but it was removed since DPI scaling is intended behavior. I'm generally against reintroducing this option, since the main problematic platform to HiDPI setups (like one 4k monitor and 1 fullhd) is `X11`, DPI settings on which already can be controlled by `WINIT_HIDPI_FACTOR=1` (@chrisduerr mentioned it in a comment above) or with `Xft.dpi` in `.Xresources`.

For example on `wayland` this setting doesn't make sense at all, since without DPI scaling your text will be blurry (not small, just blurry, and you won't be able to fix this problem with changing the font size, like you can do on X11).

**Comment by philk982** [comment]
> The only thing Alacritty cares about is the physical size and resolution of the screen. It then calculates the DPI automatically to always be correct. This allows scaling windows based on the specific DPI of each monitor, instead of the DPI of the complete X screen.

I'm experiencing the same problem. I've set the physical size of the screen in `Xorg.conf`. However, Alacritty doesn't recognize any dimensions.

    Estimated DPR: 1.75
    Estimated Dimensions: None

Where does it actually get the estimated dimensions from?

**Comment by kchibisov** [comment]
`Dimensions` is not the value you're looking for, it's the value related to `-d` flag, it's not about dimensions, the thing you're looking for is `DPR: 1.75`, which is a scale factor for your display, based on Xrandr. If it's too big, you can just execute `xrandr` and get proper dimensions.

**Comment by MahanFathi** [comment]
Simply put `WINIT_X11_SCALE_FACTOR: "1.0"` under `env:` section of your default `alacritty.yml` config.

**Comment by themaster567** [comment]
> Simply put `WINIT_X11_SCALE_FACTOR: "1.0"` under `env:` section of your default `alacritty.yml` config.
I am having the issue as all the others and this does not help.

**Comment by Pratham82** [comment]
@MahanFathi Dude thank you so much, I was tinkering with X11 for an hour but wasn't able to fix this, but your solution worked.

`
env:
  WINIT_X11_SCALE_FACTOR: '1.0'
`

**Comment by wowpetr** [comment]
I set Alacritty to launch on startup automatically in Ubuntu with Gnome and I have similar issue. The very first window that ran on startup had increased font size compared to the windows opened after that manually and only setting :
```
env:
WINIT_X11_SCALE_FACTOR: '1.0'
```
fixed that.

**Comment by tl87** [comment]
I fixed it by  adding '--scale 0.8x0.8' to my xrandr command for the screens that are active :-)

Example:

```bash
xrandr --output eDP-1 --primary --mode 1920x1280 --scale 0.8x0.8 --pos 0x0 --rotate normal
```
